library/src/main/java/com/hannesdorfmann/adapterdelegates3/DiffDelegationAdapter.java

```diff
@@ -0,0 +1,42 @@
+package com.hannesdorfmann.adapterdelegates3;
+
+import android.support.annotation.NonNull;
+import android.support.v7.util.DiffUtil;
+
+import java.util.List;
+
+/**
+ * Similar to {@link ListDelegationAdapter}, this implementation designed for items organized in a {@link List}.
+ * This adapter implementation is ready to go. All you have to do is to add {@link AdapterDelegate}s to the
+ * internal {@link AdapterDelegatesManager} i.e. in the constructor:
+ * <p>
+ * <pre>
+ * {@code
+ *    class MyAdapter extends DiffDelegationAdapter<List<Foo>> {
+ *        public MyAdapter() {
+ *            this.delegatesManager.add(new FooAdapterDelegate());
+ *            this.delegatesManager.add(new BarAdapterDelegate());
+ *        }
+ *    }
+ * }
+ * </pre>
+ *
+ * @param <T> The type of the items. Must implement {@link DiffItem}
+ * @author Sergey Opivalov
+ */
+
+public class DiffDelegationAdapter<T extends DiffItem> extends AbsDiffDelegationAdapter<T> {
+
+    public DiffDelegationAdapter() {
+        super(new DiffCallbackImpl<T>());
+    }
+
+    public DiffDelegationAdapter(@NonNull DiffUtil.ItemCallback<T> itemCallback) {
```

Comment:
If you can pass in any arbitarry `DiffUtil.ItemCallback<T>` then the restriction that it must be `DiffItem` is not useful anymore because one could write a `DiffUtil.ItemCallback` that is not using `DiffItem.getItemId()` nor `DiffItem.getItemHash()`.

As I also pointed out in some other comments, I'm not sure if this library should contain something like `DiffItem` and a default `DiffCallbackImpl` respectively.

Reply:
But T is restricted with DiffItem interface in DiffDelegationAdapter class signature.

Reply:
but no one is forced to actually use `DiffItem` api's like `DiffItem. getItemHash()`. The point I try to make is that you could pass in a DiffCallback that is not using `DiffItem` api at all, hence this restriction to T extends DiffItems gives guarantees. Since there are no guarantees, the design is flawn a bit. Thats the point I try to make.

Reply:
Yes, i've got it. So do you have any ideas about it?

Reply:
But, again, DiffItem has only two methods,so you'll not be able to implement DiffUtil.Callback with those restriction if you will not use DiffItem api, isn't it?

Reply:
Thanks for your Feedback. Maybe I'm too nitpicking but I still think that `T extends DiffItem` adds no guarantees and therefore I would like to discuss if it makes sense to have that at all.

Example how you can trick `T extends DiffItem` restriction:

```java
class Foo implements DiffItem {
    public int id;
    public String name;

     @Override public  long getItemId() { return 0; } // Return value Doesn't really matter because we don't use it anyway

     @Override int getItemHash() { return 0; } // Return value Doesn't really matter because we don't use it anyway
}
```

```java
class MyCallback extends DiffUtil.ItemCallback<Foo> {
    @Override public boolean areItemsTheSame(Foo a,  Foo b){
         return a.name.length == b.name.length;
    }
    boolean areContentTheSame(T item) {
        return a.name.equals(b.name);
    }
}
```

```java
DiffDelegationAdapter<Foo> adapter = new DiffDelegationAdapter<>( MyCallback() );
```

This compiles just fine but it actually contradicts the public API contract since `Foo.getItemId()` nor `Foo. getItemHash()` is used at all although the public API contract of `DiffItem` specifies that this two methods must be used.

That's the point I try to make. A public API should not allow this. So there are two possibility to fix this:

1. Don't allow people to pass in a custom `DiffUtil.ItemCallback`. This means that always the internally `DiffCallbackImpl`is used. By doing so, the restriction `T extends DiffItem` holds because there is no way around properly implementing `DiffItem.getItemId()` and `DiffItem.getItemHash()`  since AdapterDelegates library calls this methods for sure.
2. Remove `T extends DiffItem` as this condition adds no guarantees to the public API contract as shown in my code snippets above.  Furthermore, this means that `DiffItem` is useless at all and should removed too.

I tend to say we should go with route number 2 as I, personally, think that option number 1 is to restritive. Making AdapterDelegates make this decision for the user of the library is an necessary move in my oppinion.  I do understand you motivation to even more simplify things and remove boilerplate but I think the `DiffUtil.ItemCallback` is already simple enough so that whoever wants to use `DiffDelegationAdapter` is able to implement a `DiffUtil.ItemCallback`.

With that being said, nobody actually stops anyone defining something like `DiffItem` in your app and use `DiffItem` and the corresponding `DiffCallbackImpl` in your app. I just wouldn't put it into AdapterDelegates library.

What do you think?

Reply:
So, i do understand your thoughts. When i took a first look at AdapterDelegates, i was wondered why there is no some interface marker like DisplayableItem or smth like that =) But now i'm understanding your approach =) So, actually i'm agree with you, let's give to users possibility to implement DiffItem by themselves. But how do you think, maybe we can make some example in README?

Reply:
Yeah, that makes a ton of sense to me. Let's do it.